Adds resize mirror plugin

tsov opened the pull request:
### This PR implements

- `ResizeMirror` plugin, which resizes the mirror while dragging based on draggable element or container you are hovering over.

### Does this PR require the Docs to be updated?

Yup

### Does this PR require new tests?

Yup

### This branch been tested on...

**Browsers:**

* [x] Chrome _version_
* [x] Firefox _version_
* [x] Safari _version_
* [x] IE / Edge _version_
* [x] iOS Browser _version_
* [x] Android Browser _version_

beefchimi commented:
Going to give this a try in Examples over the weekend!

tsov commented:
@beefchimi I'll try and apply plugin on the examples pages. Want to post a GIF as well, so people see what this does 👍 

But do test it out :)

beefchimi commented on src/Plugins/ResizeMirror/ResizeMirror.js:
```diff
@@ -0,0 +1,109 @@
+
+      dragEvent.mirror.style.width = `${overRect.width}px`;
+      dragEvent.mirror.style.height = `${overRect.height}px`;
+      dragEvent.mirror.style.transition = `width ${options.duration}ms ${options.easingFunction},` +
```
This isn't going to work - you are applying the `transition` logic at the same time as the dimension change. Simply put - its not going to transition. This needs to be added before... why not just add it on `start`.

That being said - this does not take into consideration any `transition` styles already applied to that element. If you want to take this approach, you would have to `getComputedStyle` and "add on" to the list of existing `transition` properties.

In general, I'm against this kind of animation hijacking. I personally think that its better to JUST change dimensions and not apply a transition. That should be up to the consumer. That way, I can customize the easing and timing to my liking using CSS.

What do you think?

ianmart1n commented on src/Plugins/ResizeMirror/README.md:
```diff
@@ -0,0 +1,59 @@
+## ResizeMirror
+
+The resize mirror plugin changes the mirror dimensions as you drag it over other draggable elements. etc etc
```
what does "mirror" mean here? (also you can probably lose the `etc etc` 😄)

ianmart1n commented on src/Plugins/ResizeMirror/README.md:
```diff
@@ -0,0 +1,59 @@
+### API
+
+**`new ResizeMirror(draggable: Draggable): ResizeMirror`**  
+To create a resize mirror plugin instance.
```
maybe just `Creates a resize mirror plugin instance.`

ianmart1n commented on src/Plugins/ResizeMirror/README.md:
```diff
@@ -0,0 +1,59 @@
+### Options
+
+**`duration {Integer}`**  
+The duration option allows you to specify the animation duration when resizing the mirror. Default: `150`
```
maybe `Specifies the duration of the animation when resizing the mirror.`

ianmart1n commented on src/Plugins/ResizeMirror/README.md:
```diff
@@ -0,0 +1,59 @@
+The duration option allows you to specify the animation duration when resizing the mirror. Default: `150`
+
+**`easingFunction {String}`**  
+The easing option allows you to specify an animation easing function when resizing the mirror. Default: `'ease-in-out'`
```
`Specifies an easing function for the animation that runs when resizing the mirror.`

ianmart1n commented on src/Plugins/ResizeMirror/README.md:
```diff
@@ -0,0 +1,59 @@
+
+### Caveats
+
+- Does not work with empty containers (no draggable elements to base dimensions off)
```
maybe `The ResizeMirror plugin doesn't work with empty containers.` I don't know that you need to give the explanation

tsov commented:
This should be good for another round of 👀 

I removed the transitions for the mirror, which now allows the user to add their own via CSS. This also means that this plugin won't require any options

beefchimi commented on src/Plugins/ResizeMirror/tests/ResizeMirror.test.js:
```diff
@@ -0,0 +1,139 @@
+  <ul class="Container">
+    <li>Smaller item</li>
+  </ul>
+  <div class="Container">
```
Kinda wacky sample markup? Even though its a test... I can't help but feel like we either don't use `<li>` or we change the `div` to `ul`

tsov replied:
oups, that was definitely not intended... no idea what I was thinking, will change

ianmart1n commented on src/Plugins/ResizeMirror/README.md:
```diff
@@ -0,0 +1,45 @@
+It also appends the mirror element into draggable containers you drag into.
+Add transitions to the mirror element to animate the dimensions change.
+
+This plugin is not included by default, so make sure to import.
```
I wanted this to be a little more explicit, but i'm not sure what the right term is. but something like: `This plugin is not included in the default Draggable bundle, so you'll need to import it separately.` I don't know whether it should be "bundle" or "package" or what

tsov replied:
Bundle is perfectly fine 👍 a package could potentially contain multiple bundles

ianmart1n commented on src/Plugins/ResizeMirror/README.md:
```diff
@@ -0,0 +1,45 @@
+### API
+
+**`new ResizeMirror(draggable: Draggable): ResizeMirror`**  
+Creates a resize mirror plugin instance.
```
maybe `Creates an instance of the ResizeMirror plugin.`

ianmart1n commented on src/Plugins/ResizeMirror/ResizeMirror.js:
```diff
@@ -0,0 +1,93 @@
+export const defaultOptions = {};
+
+/**
+ * ResizeMirror plugin resizes the mirror element based on the draggable elements it's hovering over
```
does the mirror element's size depend on the particular draggable element that it's over? or does it change to a set size when it's over any draggable element

tsov replied:
The plugin is basically trying to find draggable elements you are hovering over and applying/copying the dimensions of the draggable element, that you are currently hovering over, onto the mirror.

tsov commented:
@ianmart1n Thanks for the reviews 🙏 I addressed all comments. Let's merge this in for the next beta release, but feel free to leave more comments, I'd put up follow-up PR to address them 👍

beefchimi commented on examples/src/content/Sortable/MultipleContainers/index.js:
```diff
@@ -17,9 +16,8 @@ export default function MultipleContainers() {
-    mirror: {
-      constrainDimensions: true,
-    },
+    // mirror: { constrainDimensions: true, },
```
@tsov is this meant to be removed if using `ResizeMirror`?

beefchimi commented on examples/src/content/Swappable/Floated/index.js:
```diff
@@ -1,5 +1,5 @@
 /* eslint-disable import/no-extraneous-dependencies, import/no-unresolved */
-import {Swappable} from '@shopify/draggable';
+import {Swappable, Plugins} from '@shopify/draggable';
```
Added to this example as it works better than the MultipleContainers example...

You may notice that the `height` in MultipleContainers changes instantly (does not transition). This is some CSS bullshit no doubt. So... I might make all of these Swappable examples use the ResizeMirror plugin.

beefchimi commented on src/Plugins/ResizeMirror/ResizeMirror.js:
```diff
@@ -0,0 +1,100 @@
+   * @param {DragOverEvent} dragEvent
+   * @private
+   */
+  [onDragOver](dragEvent) {
```
@tsov from here to the end of the file... I frigged with your code.

The issues I was experiencing before I made my changes:

1. Picking up a item would collapse its `width` and `height` _(it was unable to computer anything from what got picked up for `overElement`)_
2. Item would not get `width` + `height` until I moved `1px`.
3. I could not transition items, because the container appending was happening in the same frame as the `style` changes _(this is a no-no)_
4. Even with a single `animationFrame` separating the appending and the styling, that was not enough! I needed to double-wrap animation frames.

beefchimi replied:
Actually - the issue where the width / height isn't applied correctly on initial pickup is still present sometimes... dang!

beefchimi commented on src/Plugins/ResizeMirror/ResizeMirror.js:
```diff
@@ -0,0 +1,100 @@
+
+function animateResize(mirror, overElement) {
+  return () => {
+    const {width, height} = overElement.getBoundingClientRect();
```
I need to also use `top, left` and do some computing to get the resized item centered on the cursor.

beefchimi commented:
@tsov where did we leave off with this? Do you have uncommited work?

tsov commented:
@beefchimi has uncommited changes, will push soon and get you to 👀 Should be good to go though 👍
